lib/logstash/outputs/tcp.rb

```diff
@@ -137,67 +137,81 @@ def register
+        break if @closed.value
+        client_socket = server_socket.accept_nonblock exception: false
+        if client_socket == :wait_readable
+          IO.select [ server_socket ]
+          next
+        end
+        Thread.start(client_socket) do |client_socket|
+          # monkeypatch a 'peer' method onto the socket.
+          client_socket.instance_eval { class << self; include ::LogStash::Util::SocketPeer end }
+          @logger.debug("accepted connection", client: client_socket.peer, server: "#{@host}:#{@port}")
+          client = Client.new(client_socket, @logger)
+          Thread.current[:client] = client
+          LogStash::Util.set_thread_name("[#{pipeline_id}]|output|tcp|client_socket-#{@client_threads.size}")
+          @client_threads << Thread.current
+          client.run unless @closed.value
         end
       end
+    end
 
-      @codec.on_event do |event, payload|
-        @client_threads.select!(&:alive?)
-        @client_threads.each do |client_thread|
-          client_thread[:client].write(payload)
-        end
+    @codec.on_event do |event, payload|
+      @client_threads.select!(&:alive?)
+      @client_threads.each do |client_thread|
+        client_thread[:client].write(payload)
       end
-    else
-      client_socket = nil
-      @codec.on_event do |event, payload|
-        begin
-          client_socket = connect unless client_socket
-          r,w,e = IO.select([client_socket], [client_socket], [client_socket], nil)
-          # don't expect any reads, but a readable socket might
-          # mean the remote end closed, so read it and throw it away.
-          # we'll get an EOFError if it happens.
-          client_socket.sysread(16384) if r.any?
-
-          # Now send the payload
-          client_socket.syswrite(payload) if w.any?
-        rescue => e
-          log_warn "client socket failed:", e, host: @host, port: @port, socket: (client_socket ? client_socket.to_s : nil)
-          client_socket.close rescue nil
-          client_socket = nil
-          sleep @reconnect_interval
-          retry
+    end
+  end
+
+  def run_as_client
+    client_socket = nil
+    @codec.on_event do |event, payload|
+      begin
+        client_socket = connect unless client_socket
+        r,w,_ = IO.select([client_socket],[client_socket])
+
+        # don't expect any reads, but a readable socket might
+        # mean the remote end closed, so read it and throw it away.
+        # we'll get an EOFError if it happens.
+        r.each { |readable| readable.sysread(16384) }
+
+        # first IO#select attempt allows to be sure read socket is available
+        # rest of IO#select calls allow to wait the socket to be writable ready
+        while w == nil || w.any? == false
+          r,w,_ = IO.select([client_socket],[client_socket])
+        end
+
+        while payload && payload.bytesize > 0
+          written = client_socket.syswrite(payload)
+          payload = payload.byteslice(written..-1)
         end
```

Comment:
@yaauie , L199 to L208 the lines took from old PR and decorated. Did improve little readability by splitting the server & client side logics where in one LONG method.

Reply:
Yep, I track that. My notes thus far have been _entirely_ about pre-existing problems that really have nothing to do with this PR. I'll review your actual _effective_ changes tomorrow when my brain is more clear :)

Reply:
Thanks, I will also take a look at comments and if you think any of them are critical, I will definitely carry on. For now, as a client, current change promises to transfer entire payload with both SSL enabled & disabled options.